### ISSUE TITLE
TypeError: Cannot read property 'pods' of undefined

### ISSUE THREAD

**Issue by DazWilkin** [brendandburns/ksql#1]
If I run 'node ksql' and do nothing else, after about 8 seconds, the error results:

```
node ksql
```

```
Loading current context: "test"
> /src/ksql.js:158
    client.pods.get(function (err, pods) {
          ^

TypeError: Cannot read property 'pods' of undefined
    at load_pods (/src/ksql.js:158:11)
    at Timeout.load [as _onTimeout] (/src/ksql.js:216:2)
    at tryOnTimeout (timers.js:224:11)
    at Timer.listOnTimeout (timers.js:198:5)
```

If I'm faster, I can get a command in before the timeout:

```
node ksql
Loading current context: "test"
> select * from containers
┌──────────────────────────────────────────────────────────────────┬──────────────────────────────────────┬──────────┐
│ image                                                            │ uid                                  │ restarts │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/etcd:2.2.1                              │ 32baa2c9-2aaa-11e6-8567-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/hyperkube-amd64:v1.2.4                  │ 0b7c5d3c-2aaa-11e6-8567-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/hyperkube-amd64:v1.2.4                  │ 0b7c5d3c-2aaa-11e6-8567-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/hyperkube-amd64:v1.2.4                  │ 0b7c5d3c-2aaa-11e6-8567-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/hyperkube-amd64:v1.2.4                  │ 0b7c5d3c-2aaa-11e6-8567-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/hyperkube-amd64:v1.2.4                  │ 39491447-2aaa-11e6-8567-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ nginx                                                            │ 6266eae8-2b6e-11e6-b050-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ nginx                                                            │ 61529733-2b6e-11e6-b050-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ nginx                                                            │ 6266d66e-2b6e-11e6-b050-001fc69b9496 │ 0        │
├──────────────────────────────────────────────────────────────────┼──────────────────────────────────────┼──────────┤
│ gcr.io/google_containers/kubernetes-dashboard-amd64:v1.1.0-beta2 │ 8e49617a-2b6e-11e6-b050-001fc69b9496 │ 0        │
└──────────────────────────────────────────────────────────────────┴──────────────────────────────────────┴──────────┘
> /src/ksql.js:158
    client.pods.get(function (err, pods) {
          ^

TypeError: Cannot read property 'pods' of undefined
    at load_pods (/src/ksql.js:158:11)
    at Timeout.load [as _onTimeout] (/src/ksql.js:216:2)
    at tryOnTimeout (timers.js:224:11)
    at Timer.listOnTimeout (timers.js:198:5)
```

I'm running Kubernetes locally:
http://kubernetes.io/docs/getting-started-guides/docker/

and (!) attempting to run KSQL from a container:

added package.json:

```
{
  "name": "KSQL",
  "description": "KSQL - A simple tool for interactive database queries on Kubernetes resources.",
  "main": "ksql.js",
  "scripts": {
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/brendandburns/ksql.git"
  },
  "author": "Brendan Burns",
  "bugs": {
    "url": "https://github.com/brendandburns/ksql/issues"
  },
  "homepage": "https://github.com/brendandburns/ksql#readme",
  "dependencies": {
    "alasql": "^0.2.7",
    "cli-table2": "^0.2.0",
    "js-yaml": "^3.6.1",
    "node-kubernetes-client": "^0.2.3",
    "q": "^1.4.1",
    "readline-history": "^1.2.0"
  }
}
```

added Dockerfile:

```
FROM node

ENV DIRPATH /src
RUN mkdir $DIRPATH 
WORKDIR $DIRPATH

ADD package.json .
ADD index.html .
ADD ksql.js .
ADD script.js .
ADD style.css .

RUN npm install

# install kubectl
RUN wget https://storage.googleapis.com/kubernetes-release/release/v1.0.1/bin/linux/amd64/kubectl
RUN chmod +x kubectl
RUN mv kubectl /usr/local/bin

# running Kubernetes locally
RUN kubectl config set-cluster test --server=http://localhost:8080
RUN kubectl config set-context test --cluster=test
RUN kubectl config use-context test

RUN kubectl proxy &
```

and

```
docker build --tag=ksql .
docker run -it --net="host" dazwilkin/ksql /bin/bash
```

**Comment by DazWilkin** [comment]
I think the issue is a result of either the v1 API or the cluster that I'm running.

Your code expects a pod enumerator:

https://github.com/brendandburns/ksql/blob/master/ksql.js#L158

```
client.pods.get(function (err, pods) { ... }
```

But this appears to no longer exist in node-kubernetes-client

https://github.com/tenxcloud/node-kubernetes-client/blob/master/test/test-pods.js#L41

```
client.pods.get(pods[0].id, function (err, pod) { ... }
```

But node-kubernetes-client code is incorrect, in the test, it is passed pods[0].id but pods is set here:

```
client.pods.getBy({"namespace": "default"}, function (err, podsArr) {
...
pods = podsArr.items;
...
}
```

and the JSON for an item is

```
{
  metadata: {
    uid: ... ,
    ... 
  }
  spec: ... ,
  status: ...
}
```

and metadata includes uid which is, I think what should be passed (and what your code expects).

However, the v1 API does not appear to support a method to return details for a specific pod. And, if I try to forcibly call the node-kubernetes-client client.pods.get using a UID, I get an error:

```
pod:  undefined
{ [Error: [object Object]]
  message: 
   { kind: 'Status',
     apiVersion: 'v1',
     metadata: {},
     status: 'Failure',
     message: 'the server could not find the requested resource',
     reason: 'NotFound',
     details: {},
     code: 404 },
  statusCode: 404 }
```

Here's my code:

```
var Client = require('node-kubernetes-client');
var q = require('q');

var client = new Client({
    host:  '127.0.0.1:8080',
    protocol: 'http',
    version: 'v1'
});

// The UID of one of my pods returned by 'getBy' -- avoid 'get' before 'getBy'
var pods = [
  {"metadata":
    {"uid":"92b7ac61-2b74-11e6-b050-001fc69b9496"}
  }
];

// This works
client.pods.getBy({"namespace": "default"}, function(err, podsArr) {
  console.log(err);
  pods = podsArr.items;
  console.log('pods: '+ JSON.stringify(pods));
});

// This yields a 404
client.pods.get(pods[0].metadata.uid, function (err, pod) {
  console.log(err);
  console.log('pod:', pod);
});
```

**Comment by brendandburns** [comment]
Thanks, I will try to repro this and see what I can do to fix.

**Comment by brendandburns** [comment]
Ok, sorry for the delay.  This was actually due to an error where the client was getting reset to null on the polling.  This should now be fixed, see #2
